### PR TITLE
Tests and completion updates

### LINKED ISSUE #107: Build should include liberty-ls
The current Github actions are only building lemminx-liberty and not liberty-ls. We should build both when PRs are created/modified.

### PR DESCRIPTION
**Pull request by evie-lau**
Some work for #89 and resolves #107 

- Modify Github Actions to run the build for liberty-ls
- Add tests for hover and completion features for liberty-ls
- Modify completion behavior a bit to filter results on language server-side, instead of relying completely on client-side automatic filtering.

### CODE REVIEW

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java [hunk 1]
```diff
@@ -23,13 +23,15 @@
...
 import org.eclipse.lsp4j.TextDocumentSyncKind;
 import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.lsp4j.services.LanguageClientAware;
 import org.eclipse.lsp4j.services.LanguageServer;
 import org.eclipse.lsp4j.services.TextDocumentService;
 import org.eclipse.lsp4j.services.WorkspaceService;
 
 
-public class LibertyLanguageServer implements LanguageServer {
+public class LibertyLanguageServer extends AbstractLanguageServer implements LanguageServer, LanguageClientAware {
```
Comment (cherylking) [inline liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java:32]: Will these changes affect how the various IDEs start/interact with the Liberty Language Server?

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java [hunk 2]
```diff
@@ -76,14 +78,20 @@ public synchronized void updateSettings(Object initializationOptionsSettings) {
         // TODO: else update settings
     }
 
+    @Override
```
Comment (cherylking) [inline liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java:81]: Are these really overrides? I don't see method connect, shutdown or exit in the class that is being extended. I see these were already overrides before, but I think that is incorrect. Also, is there a need to use `super` when the method is not in the current class and is inherited?

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/Messages.java [hunk 3]
```diff
@@ -80,10 +82,22 @@ public static List<String> getMatchingKeys(String query, LibertyTextDocument tex
         }
         
         String filename = textDocument.getUri();
+        // remove completion results that don't contain the query string (case-insensitive search)
```
Comment (cherylking) [inline liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/Messages.java:85]: I don't really follow what this is for. I read the comment, but that is not telling me what the query string is for (example of something we are removing). What is the use case?

**Review by cherylking — CHANGES_REQUESTED** [review]

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/Messages.java (hunk 3 above)
Comment (evie-lau) [reply liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/Messages.java:85]: Some background... the initial implementation never filtered any results for completion, just returned the entire catalog of keys, and let the client-side IDE handle the filtering based on what was typed. 
When it came to running tests, each completion request would return that full list, making the tests pretty useless other than checking that: yes, we return all the possible keys for this file.

Comment (evie-lau) [reply liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/Messages.java:85]: Functionally when running the IDE extension with the language server, it doesn't seem to make much of a difference. But adding this logic to filter the results on the server-side made it a lot easier to test reasonable functionality for the completion feature.

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java (hunk 1 above)
Comment (evie-lau) [reply liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java:32]: I don't think they will change. They should still use the original initialization methods we had in place before, and these additions are in place for our test suite to startup the language server.

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java (hunk 2 above)
Comment (evie-lau) [reply liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java:81]: Hmm, you are right, they're not overriding any superclass methods, but these methods are implementing the interface methods.
I think it is still a good marker to show it is an implemented method, even if it's not necessary. (I also read that Eclipse auto-marks implemented methods with `@Override` with Java 1.6+)
`super` is also not necessary, and serves more as an indicator. Should I remove these?

Comment (cherylking) [reply liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java:81]: k...if Eclipse auto-marks them that is fine. Just leave it.

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/Messages.java (hunk 3 above)
Comment (cherylking) [reply liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/Messages.java:85]: So it filters out based on a key? Or it is filtering out bootstrap vs server.env? Still not clear.

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java (hunk 1 above)
Comment (cherylking) [reply liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java:32]: The functionality looks different in the original initialization methods. That is why I asked this question.

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java [hunk 4]
```diff
@@ -76,14 +78,20 @@ public synchronized void updateSettings(Object initializationOptionsSettings) {
...
+    public void connect(LanguageClient client) {
+        this.languageClient = client;
+    }
+
     @Override
     public CompletableFuture<Object> shutdown() {
-        // TODO Auto-generated method stub
-        return null;
+        super.shutdownServer();
+        return CompletableFuture.completedFuture(new Object());
```
Comment (cherylking) [inline liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java:89]: The contents of the shutdown and exit method look pretty different to me from before. This is why I thought behavior might be different?

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/Messages.java (hunk 3 above)
Comment (evie-lau) [reply liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/Messages.java:85]: Yes, it filters the available keys based on the entered text (query).

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java (hunk 1 above)
Comment (evie-lau) [reply liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java:32]: You may be looking at the initialization for the Test language server?

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java (hunk 4 above)
Comment (evie-lau) [reply liberty-ls/src/main/java/io/openliberty/tools/langserver/LibertyLanguageServer.java:89]: Ah I see. The updated methods are just to ensure the language server instances from the tests are closed out.

**Review by cherylking — APPROVED** [review]

**Pull request merged**